src/easylink/step.py

```diff
@@ -261,6 +277,122 @@ def update_implementation_graph(
+    ) -> None:
+        super(CompositeStep, self).__init__(step_name, name, input_slots, output_slots)
+        # TODO [MIC-5135]: Make loopstep compatible with sequence of steps using step hierarchy (composite steps)
+        if len(nodes) != 1 or nodes[0].name != step_name:
+            raise NotImplementedError(
+                f"LoopStep {self.name} must be initialized with a single node with the same name."
+            )
+        if not isinstance(nodes[0], BasicStep):
+            raise NotImplementedError(
+                f"LoopStep {self.name} can currently only loop a single implementation."
+            )
+        self.unlooped_node = nodes[0]
+        for edge in edges:
+            if not edge.in_node == edge.out_node == step_name:
+                raise NotImplementedError(
+                    f"LoopStep {self.name} must be initialized with only self-loops as edges"
+                )
+        self.edges = edges
+
+    @property
+    def config_key(self):
+        return "iterate"
+
+    def update_implementation_graph(
+        self, graph: nx.MultiDiGraph, step_config: LayeredConfigTree
+    ) -> None:
+        if not self.name in step_config or not self.config_key in step_config[self.name]:
+            BasicStep.update_implementation_graph(self, graph, step_config)
         else:
             sub_config = step_config[self.name][self.config_key]
-            return CompositeStep.validate_step(self, sub_config)
+            num_loops = len(sub_config)
+            self.graph = self._create_looped_graph(num_loops)
+            self.slot_mappings = self.get_loop_slot_mappings(num_loops)
+            loop_config = self.get_loop_config(sub_config)
+            CompositeStep.update_implementation_graph(self, graph, loop_config)
+
+    def validate_step(self, step_config: LayeredConfigTree) -> Dict[str, List[str]]:
+        if not self.name in step_config or not self.config_key in step_config[self.name]:
+            return BasicStep.validate_step(self, step_config)
+
+        sub_config = step_config[self.name][self.config_key]
+
+        if not isinstance(sub_config, list):
+            return {
+                f"step {self.name}": [
+                    "Loops must be formatted as a sequence in the pipeline configuration."
+                ]
+            }
+
+        if len(sub_config) == 0:
+            return {f"step {self.name}": ["No loops configured under iterate key."]}
+
+        errors = {}
+        for i, loop in enumerate(sub_config):
+            loop_errors = self.unlooped_node.validate_step({self.name: loop})
+            if loop_errors:
+                errors[f"step {self.name}"][f"loop {i+1}"] = loop_errors
+        return errors
+
+    def _create_looped_graph(self, num_loops: int) -> nx.MultiDiGraph:
+        graph = nx.MultiDiGraph()
+
+        for i in range(num_loops):
+            updated_step = copy.deepcopy(self.unlooped_node)
```

Comment:
Is this a step? Or an implementation?

Reply:
`self.unlooped_node` is a step. We're basically "expanding" the Step subgraph before grabbing the implementations from it. Since we need the config to know how many loops to make, I'm doing it in `update_implementation_graph` but I considered breaking this into a separate "phase" between validation and getting the implementations which any type of step could concevably make use of (since multiples will have the same sort of thing happening)

Reply:
> I considered breaking this into a separate "phase" between validation and getting the implementations which any type of step could concevably make use of (since multiples will have the same sort of thing happening)

Interesting idea, I'd keep it in mind as you go to see if it would simplify things!